client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,489 @@
+	// 1) To read (and set) the spenderBlock field.
+	// 2) To prevent duplicate spender block scans if the spenderBlock is not
+	//    already known. Holding this lock now ensures that any ongoing scan
+	//    completes before we try to access the output.spenderBlock field
+	//    which may prevent unnecessary rescan.
+	output.spenderMtx.Lock()
+	defer output.spenderMtx.Unlock()
+
+	// Check if this output is known to be spent in a mainchain block.
+	if output.spenderBlock != nil {
+		spenderBlockStillValid, err := dcr.isMainchainBlock(ctx, output.spenderBlock)
+		if err != nil {
+			return false, err
+		}
+		if spenderBlockStillValid {
+			dcr.log.Debugf("Found cached information for the spender of %s.", output.op)
+			return true, nil
+		}
+		// Output was previously found to have been spent but the block
+		// containing the spending tx seems to have been invalidated.
+		dcr.log.Warnf("Block %s found to contain spender of output %s has been invalidated.",
+			output.spenderBlock.hash, output.op)
+		output.spenderBlock = nil
+	}
+
+	// This tx output is not known to be spent as of last search (if any).
+	// Scan block filters starting from the block after the tx block or the
+	// lastScannedBlock (if there was a previous scan). Use mainchainAncestor
+	// to ensure that scanning starts from a mainchain block in the event that
+	// the lastScannedBlock have been re-orged out of the mainchain. We already
+	// checked that the txBlock is not invalidated above.
+	_, lastScannedHeight, err := dcr.mainchainAncestor(ctx, output.lastScannedBlock)
+	if err != nil {
+		return false, err
+	}
+	nextScanHeight := lastScannedHeight + 1
+
+	bestBlock := dcr.cachedBestBlock()
+	if nextScanHeight >= bestBlock.height {
+		if nextScanHeight > bestBlock.height {
+			dcr.log.Warnf("Attempted to look for output spender in block %d but current tip is %d!",
+				nextScanHeight, bestBlock.height)
+		}
+		// No new blocks to scan, output isn't spent as of last scan.
+		return false, nil
+	}
+
+	// Search for this output's spender in the blocks between startBlock and
+	// the current best block.
+	nextScanHash, err := dcr.wallet.GetBlockHash(ctx, nextScanHeight)
+	if err != nil {
+		return false, err
+	}
+	spenderTx, stopBlockHash, err := dcr.findTxOutSpender(ctx, output.op, output.PkScript, &block{nextScanHeight, nextScanHash})
+	if stopBlockHash != nil { // might be nil if the search never scanned a block
+		output.lastScannedBlock = stopBlockHash
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// Cache relevant spender info if the spender is found.
+	spent := spenderTx != nil
+	if spent {
+		spenderBlockHash, err := chainhash.NewHashFromStr(spenderTx.BlockHash)
+		if err != nil {
+			// dcr.log.Errorf("Invalid hash (%s) for tx that spends output %s: %v", spenderTx.BlockHash, output.op, err)
+			return false, fmt.Errorf("invalid hash (%s) for tx that spends output %s: %v", spenderTx.BlockHash, output.op, err)
+		} else {
```

Comment:
If you need to do another commit, please remove the else block and commented log.

Reply:
That's the intent, yes. Your version is clearer, will just adopt it.